src/libraries/System.Private.CoreLib/src/System/Threading/Volatile.cs

```diff
@@ -13,176 +13,202 @@ public static class Volatile
-        #endregion
-
-        #region Byte
-        private struct VolatileByte { public volatile byte Value; }
+        public static void Write(ref bool location, bool value)
+        {
+            WriteBarrier();
+            location = value;
+        }
 
         [Intrinsic]
         [NonVersionable]
-        public static byte Read(ref readonly byte location) =>
-            Unsafe.As<byte, VolatileByte>(ref Unsafe.AsRef(in location)).Value;
+        public static byte Read(ref readonly byte location)
+        {
+            byte value = location;
+            ReadBarrier();
+            return value;
+        }
 
         [Intrinsic]
         [NonVersionable]
-        public static void Write(ref byte location, byte value) =>
-            Unsafe.As<byte, VolatileByte>(ref location).Value = value;
-        #endregion
+        public static void Write(ref byte location, byte value)
+        {
+            WriteBarrier();
+            location = value;
+        }
 
-        #region Double
         [Intrinsic]
         [NonVersionable]
-        public static double Read(ref readonly double location)
-        {
-            long result = Read(ref Unsafe.As<double, long>(ref Unsafe.AsRef(in location)));
-            return BitConverter.Int64BitsToDouble(result);
-        }
+        public static double Read(ref readonly double location) =>
+            BitConverter.Int64BitsToDouble(Read(ref Unsafe.BitCast<double, long>(location)));
 
         [Intrinsic]
         [NonVersionable]
         public static void Write(ref double location, double value) =>
             Write(ref Unsafe.As<double, long>(ref location), BitConverter.DoubleToInt64Bits(value));
-        #endregion
-
-        #region Int16
-        private struct VolatileInt16 { public volatile short Value; }
 
         [Intrinsic]
         [NonVersionable]
-        public static short Read(ref readonly short location) =>
-            Unsafe.As<short, VolatileInt16>(ref Unsafe.AsRef(in location)).Value;
+        public static short Read(ref readonly short location)
+        {
+            short value = location;
+            ReadBarrier();
+            return value;
+        }
 
         [Intrinsic]
         [NonVersionable]
-        public static void Write(ref short location, short value) =>
-            Unsafe.As<short, VolatileInt16>(ref location).Value = value;
-        #endregion
-
-        #region Int32
-        private struct VolatileInt32 { public volatile int Value; }
+        public static void Write(ref short location, short value)
+        {
+            WriteBarrier();
+            location = value;
+        }
 
         [Intrinsic]
         [NonVersionable]
-        public static int Read(ref readonly int location) =>
-            Unsafe.As<int, VolatileInt32>(ref Unsafe.AsRef(in location)).Value;
+        public static int Read(ref readonly int location)
+        {
+            int value = location;
```

Comment:
Won't this behave differently on some platforms with an unaligned ref?

Reply:
Behavior of unaligned refs is undefined 

https://github.com/dotnet/runtime/blob/main/docs/design/specs/Memory-model.md#atomic-memory-accesses